text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,242 @@
+   [stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable).
+
+## Terminology
+
+To leverage existing standards, the terminology used in this document is based
+on the [CloudEvents specification](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md).
+CloudEvents is hosted by the CNCF and provides a specification for describing
+event data in common formats to provide interoperability across services,
+platforms and systems.
+
+### Message
+
+A "message" is a transport envelope for the transfer of information. The
+information is a combination of a payload and metadata.  Messages may be
+transferred directly between parties or via one or more intermediaries.  A
+message may carry annotations that are directed at intermediaries on the
+message path.  Messages are uniquely identifiable.
+
+In the strict sense, a _message_ is a payload that is sent to a specific
+destination, whereas an _event_ is a signal emitted by a component upon
+reaching a given state. This document is agnostic of those differences and uses
+the term "message" in a wider sense to cover both concepts.
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates and
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary or consumer.
+
+### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new events.
+
+The consumer receives, processes, and settles a message. "Receiving" is the
+process of obtaining a message from the intermediary, "processing" is the
+process of acting on the information a message contains, "settling" is the
+process of notifying an intermediary that a message was processed successfully.
+
+### Intermediary
+
+An "intermediary" receives a message to forward it to the next receiver, which
+might be another intermediary or a consumer.
+
+## Scenarios
+
+Producing and consuming a message involves five stages:
+
+```
+PRODUCER
+
+Create
+  |                            CONSUMER
+  v        +--------------+                   
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+       |
+                  ^               v
+                  .            Process
+                  .               |
+                  .               v
+                  . . . . . .  Settle
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In exceptional cases (fire-and-forget), the
+   settlement stage does not exist.
+
+The messaging semantic conventions need to define how to model those stages in
```

Comment:
Eventually it will be nice to further augment this with real world flows perhaps through cloud services. From what I understand, messaging can behave vastly different for scenarios like S3 - SQS - Lambda vs Kinesis vs DynamoDB streams, most importantly whether batching happens or not. It'll be good to knock out as many of these real world flows as possible and detail out user expectations for each.

Reply:
> It'll be good to knock out as many of these real world flows as possible and detail out user expectations for each.

Agreed. We expect quite some prototyping and verification work once we start to work through the conventions. Any help in doing this for SQS and Kinesis would be very welcome ...